### PR TITLE
Implement bugfix for chart dependencies and gitops.

### PR DESCRIPTION
**Pull request by komish**
~Do Not Merge.~

~I will remove :point_up: if and when this this is valid to merge/review. I have intentionally not applied the *draft* label because this prevents certain CI operations from running.~

[EDIT] This is ready for review.

---

This PR contains all of the impactful changes to how the certification pipeline works in order to resolve issues when depending on charts we provide.

In effect, this PR enables:

- Locked helm chart name tracking and generation
- Checks on OWNERS file submission to ensure locks are respected
- Fixes indexing strategy to index on chart names only

In addition, this PR includes E2E testing extensions for some of the new logic.

---
Change summary by file:

.github/workflows/mercury_bot.yml
- Only run this on changes to partner OWNERS files.
- Make sure that the incoming content is valid.
- Make sure the incoming chart name is not locked.

scripts/src/checkprcontent/checkpr.py
- Make sure we're adding the chart name to the github output.
- Ensure redhat charts follow proper standards at chart content check time. This is a secondary check to the one that's executed at OWNERS file submission time via GHA.
- Ensure the index key is set to the chart's name (changed from `org-chart`).

tests/functional/behave_features/common/utils/chart_certification.py
- Ensure the index key is set to the chart's name (changed from `org-chart`).

tests/functional/behave_features/environment.py
- Defines the OWNERS file testing fixture
- Adds the fixture to the scenario when the proper references are added in the scenario tests

.github/workflows/owners-redhat.yml (new)
- On Red Hat OWNERS file submissions, will check to make sure that the content matches standards.

.github/workflows/lock-sanity-check.yml (new)
- Sanity checks chart lock workflows by running on every main push, and notifying on failures.

.github/workflows/check-locks-on-owners-submission.yml (new)
- Ensures new redhat/community owners files do not encroach on locks.

tests/functional/behave_features/HC-20_owners_file.feature (new)
- Defines the new Helm Cert E2E test for OWNERS files,

tests/functional/behave_features/steps/implementation_owners.py (new)
- Implements the new Helm Cert E2E test for OWNERS files.

### CODE REVIEW

File: scripts/src/checkprcontent/checkpr.py [hunk 1]
```diff
@@ -207,6 +208,20 @@ def ensure_only_chart_is_modified(api_url, repository, branch):
...
+            if not chart.startswith("redhat-"):
+                msg = f"[ERROR] Charts provided by Red Hat must have their name begin with the redhat- prefix. I.e. redhat-{chart}"
+                gitutils.add_output("pr-content-error-message", msg)
+                sys.exit(1)
+
+        # Non Red Hat charts must not carry the Red Hat prefix.
+        if organization != "redhat" and chart.startswith("redhat-"):
+            msg = f"[ERROR] The redhat- prefix is reserved for charts provided by Red Hat. Your chart: {chart}"
+            gitutils.add_output("pr-content-error-message", msg)
+            sys.exit(1)
```
Comment (mgoerens) [inline scripts/src/checkprcontent/checkpr.py:222]: We are also enforcing charts under community/redhat to have this prefix, correct ?Seeing the conditions under which we run the .github/workflows/owners-redhat.yml workflow:
```
on:
  pull_request:
    paths:
    - charts/redhat/**/OWNERS
    - charts/community/redhat/**/OWNERS
```

If yes, this is going to fail here.

**Review by mgoerens — CHANGES_REQUESTED** [review]
Only blocking issue for me is the question of whether we also enforce the `redhat-` prefix on community/redhat charts. Otherwise minor comments.

Reg. action versions, there are multiple occurrences of slightly outdated actions, so if you want to fix them all at once (not optimized):
```
ACTION_NAME="actions/checkout"; grep "$ACTION_NAME" -rnl .github/ | xargs sed -i "s:$ACTION_NAME@v3:$ACTION_NAME@v4:g"
ACTION_NAME="actions/github-script"; grep "$ACTION_NAME" -rnl .github/ | xargs sed -i "s:$ACTION_NAME@v6:$ACTION_NAME@v7:g"
ACTION_NAME="actions/setup-python"; grep "$ACTION_NAME" -rnl .github/ | xargs sed -i "s:$ACTION_NAME@v4:$ACTION_NAME@v5:g"
```

Finally a general comment regarding newlines between steps. Sometimes there aren't (e.g. .github/workflows/owners-redhat.yml), sometimes there are. I feel it's more readable with newlines, but you might have a different opinion ?

File: scripts/src/checkprcontent/checkpr.py (hunk 1 above)
Comment (jsm84) [reply scripts/src/checkprcontent/checkpr.py:222]: @mgoerens we are only enforcing Red hat provided charts to have the `redhat-` prefix, while the `<vendor>-` prefix is being dropped for community and partner charts.

The reason was we didn't want Red Hat "squatting" on chart names for a community project or a partner's product (the latter case is much less likely).

File: .github/workflows/check-locks-on-owners-submission.yml [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+  pull_request_target:
+    types: [opened, synchronize, reopened, edited, ready_for_review, labeled]
+    paths:
+    - charts/community/**/OWNERS
+    - charts/redhat/**/OWNERS
+
+jobs:
+  owners-file-check:
+    name: OWNERS file PR checker
+    runs-on: ubuntu-20.04
```
[the remark this answers is not in the document]
Comment (komish) [reply .github/workflows/check-locks-on-owners-submission.yml:25]: I think making it a conscious bump avoids surprises. That said, I see both 22.04 and 20.04 being used throughout (other than latest). For the files I'm modifying/updating, I'll explicitly use 22.04 and we can update any 20.04's to 22.04 in a future patch.

File: .github/workflows/lock-sanity-check.yml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+      run: |
+        gh issue create --title "Chart lock generation sanity check is failing" --body "${{ env.COMMENT_BODY }}"
+      env:
+        GITHUB_TOKEN: ${{ secrets.BOT_TOKEN }}
+        COMMENT_BODY: |
+          While performing a sanity check of the chart generation logic on the main branch, a failure occurred!
+          
+          See workflow: ${{github.server_url}}/${{github.repository}}/actions/runs/${{github.run_id}}
+
+          cc @komish
```
[the remark this answers is not in the document]
Comment (komish) [reply .github/workflows/lock-sanity-check.yml:57]: Eventually, I'd like to cc a group, but we're not there yet.

File: .github/workflows/check-locks-on-owners-submission.yml [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+      - name: Set up Python 3.x Part 2
+        run: |
+          # set up python
+          python3 -m venv ve1
+          cd scripts
+          ../ve1/bin/pip3 install -r requirements.txt
+          ../ve1/bin/pip3 install .
+          cd ..
+
+      - name: get files changed
```
[the remark this answers is not in the document]
Comment (komish) [reply .github/workflows/check-locks-on-owners-submission.yml:46]: Agreed - I think your work on the precheck idea would be perfect to reuse here. I think when that's ready and merged in, we can rework this piece here to use it and standardize into one place.

With that in mind, I'm not seeing an action to take today. I'll let you mark this as resolved if you agree.

File: .github/workflows/check-locks-on-owners-submission.yml [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        working-directory: "pr-branch"
+        id: fact-check
+        run: |
+          file=$(echo ${{ steps.get_files_changed.outputs.pr_files }} | yq .0)
+          owner_contents_chart_name=$(yq '.chart.name' ${file})
+          owner_contents_vendor_label=$(yq '.vendor.label' ${file})
+          echo "Chart Name Comparison: ${owner_contents_chart_name} = ${{ steps.gather-metadata.outputs.chart-name }}" 
+          echo "Vendor Label Comparison: ${owner_contents_vendor_label} = ${{ steps.gather-metadata.outputs.organization }}"
+          test "${owner_contents_chart_name}" = "${{ steps.gather-metadata.outputs.chart-name }}"
+          test "${owner_contents_vendor_label}" = "${{ steps.gather-metadata.outputs.organization }}"
```
[the remark this answers is not in the document]
Comment (komish) [reply .github/workflows/check-locks-on-owners-submission.yml:131]: On one hand, I hesitate to add too many commenting "branches" to the workflow(s). On the other, I much prefer the branches to some of the abstractions we have, whereby any task can include data to be added to the PR as a comment, and they're all concatenated.

With that in mind, I've added a comment step for this step.

File: scripts/src/checkprcontent/checkpr.py (hunk 1 above)
Comment (komish) [reply scripts/src/checkprcontent/checkpr.py:222]: This is one of the earliest modifications I made, so please cross-check me intently! 

These code changes in checkpr check the **organization**, but not the **category**. That is, a chart that resides in `charts/redhat/redhat/myapp` has a category of redhat, an organization of redhat, and a chartname of myapp. 

A chart residing in `charts/community/redhat/anotherapp` has a category of community, an organization of redhat, and a chartname of anotherapp. 

The highlighted code snippet checks for **organization** values that are not redhat, and prevents them from using the prefix. Equally the workflow you mentioned (and earlier in checkpr) double checks to make sure submitted charts comply with the requirement to have the redhat prefix.

With that in mind, I believe the checkpr code is valid as written - but as mentioned, it's one of the earlier code snippets I wrote in the PoC, so please triple check me.

**Comment by komish** [comment]
@mgoerens I believe I've addressed everything you've requested.

Re: Newlines between steps. I've received that feedback before. I don't find it strictly necessary, but I also don't mind it at all. If I've missed them, I'm happy to add them. If it doesn't happen in this PR, I can go back and normalize all workflows at the same time that I'm standardizing on a single github runner image (e.g. the ubuntu-22.04 change).

File: scripts/src/checkprcontent/checkpr.py (hunk 1 above)
Comment (mgoerens) [reply scripts/src/checkprcontent/checkpr.py:222]: > These code changes in checkpr check the organization, but not the category

I overlooked that. Thanks, that answers it then.

File: .github/workflows/check-locks-on-owners-submission.yml [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+jobs:
+  owners-file-check:
+    name: OWNERS file PR checker
+    runs-on: ubuntu-20.04
+    if: |
+      github.event.pull_request.draft == false
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
```
[the remark this answers is not in the document]
Comment (mgoerens) [reply .github/workflows/check-locks-on-owners-submission.yml:30]: I still see `v3` here ? Again, nothing major, happy to take this point outside of this PR if we want to move fast.

**Review by mgoerens — APPROVED** [review]

**Pull request merged**